sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/RateLimitPolicyFactory.java

```diff
@@ -0,0 +1,105 @@
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.io.kinesis;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.List;
+import org.apache.beam.sdk.util.BackOff;
+import org.apache.beam.sdk.util.FluentBackoff;
+import org.joda.time.Duration;
+
+/**
+ * Implement this interface to create a {@code RateLimitPolicy}. Used to create a rate limiter for
+ * each shard.
+ */
+public interface RateLimitPolicyFactory extends Serializable {
+
+  RateLimitPolicy getRateLimitPolicy();
+
+  static RateLimitPolicyFactory withoutLimiter() {
+    return () -> new RateLimitPolicy() {};
+  }
+
+  static RateLimitPolicyFactory withBackoff(FluentBackoff fluentBackoff) {
+    return () -> new BackoffRateLimiter(fluentBackoff);
+  }
+
+  static RateLimitPolicyFactory withFixedDelay() {
+    return FixedDelayRateLimiter::new;
+  }
+
+  static RateLimitPolicyFactory withFixedDelay(Duration delay) {
+    return () -> new FixedDelayRateLimiter(delay);
+  }
+
+  class BackoffRateLimiter implements RateLimitPolicy {
+
+    private final BackOff backoff;
+
+    public BackoffRateLimiter(FluentBackoff fluentBackoff) {
+      this.backoff =
+          fluentBackoff
+              // never stop retrying
+              .withMaxRetries(Integer.MAX_VALUE)
+              .withMaxCumulativeBackoff(Duration.standardDays(1000))
+              .backoff();
+    }
+
+    @Override
+    public void onThrottle(KinesisClientThrottledException t) throws InterruptedException {
+      try {
+        long backOffMillis = backoff.nextBackOffMillis();
+        if (backOffMillis != BackOff.STOP) {
+          Thread.sleep(backOffMillis);
+        }
+      } catch (IOException e) {
+        // do nothing
+      }
+    }
+
+    @Override
+    public void onSuccess(List<KinesisRecord> records) {
+      try {
+        backoff.reset();
+      } catch (IOException e) {
+        // do nothing
+      }
+    }
+  }
+
+  class FixedDelayRateLimiter implements RateLimitPolicy {
+
+    private static final Duration DEFAULT_DELAY = Duration.standardSeconds(1);
+
+    private final Duration delay;
+
+    public FixedDelayRateLimiter() {
+      this(DEFAULT_DELAY);
+    }
+
+    public FixedDelayRateLimiter(Duration delay) {
+      this.delay = delay;
+    }
+
+    @Override
+    public void onSuccess(List<KinesisRecord> records) throws InterruptedException {
```

Comment:
records param is not being used?

Reply:
It's hard to anticipate what state a custom rate limit policy will want to keep but since we have the list of records and it might want them I just passed them in here. Same with passing the KinesisClientThrottledException in onThrottled(). Currently, no, there are no implementations using these.